### PR TITLE
Building a Docker Image for Deployment

### LINKED ISSUE #303: [Deployment] Add ability to dynamically generate a ASP.NET Core Docker Image
**Is your feature request related to a problem? Please describe.**
We need to generate an ASP.NET Core Web API that's dynamic and takes a `ModelInput`, `ModelOutput` and a model.zip.
These three instances will be dynamically copied over based on the deployment, while the rest stays the same.

We need the ability to dynamically generate an image with these on demand, provided #302 is completed. 

**Depends** on #302 
**Pre-req** for #304 

**Describe the solution you'd like**
To be discussed.

- Can we generate a base image with the Web API on it and then copy over the remaining artifacts and re-build/re-publish?
- Can that base image be stored in Docker Hub? 
- What should the naming of the image be? Maybe `ExperimentName` with the tag being the `DeploymentId` Guid?

### PR DESCRIPTION
**Pull request by aslotte**
## Resolves
Fixes #303 
Partially #304 

## Description
Generates a docker image for a model and copies over the model.zip

### CODE REVIEW

**Comment by aslotte** [comment]
Next steps are:
- Force docker registry running in container to use password and username
- Add these to integration test
- Run docker login on `docker push`
- Assert that an image is pushed to a repo
- Spin up a service container for these tests in GitHub
- Add more integration and unit tests

**Comment by aslotte** [comment]
@Brett-Parker  / @AnoojNair  / @willvelida - I realized I started going into #304  so I' going to stop this PR here so it doesn't get too big. #303 should be covered here. There are a couple of outstanding issues that will be added once support is there which is e.g. copying over the schema.

For #304 which will be a separate PR, what remains is `docker login` before `docker push` and more integration tests (e.g. we could start the pushed container and make sure a prediction works with a dummy model or something).

Anyhow this PR is ready for review as is!

**Comment by aslotte** [comment]
Okay, I'm hands-off on this PR now... 😄

File: src/MLOps.NET/Docker/Interfaces/ICliExecutor.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+        /// <summary>
+        /// Runs docker pull
+        /// </summary>
+        /// <param name="tagName"></param>
+        /// <returns>True if image exists</returns>
+        Task<bool> RunDockerPull(string tagName);
+
+        /// <summary>
+        /// Runs docker image rm
```
Comment (Brett-Parker) [inline src/MLOps.NET/Docker/Interfaces/ICliExecutor.cs:50]: Can this be expanded. Not clear what rm is.

File: src/MLOps.NET/Exceptions/DockerRemoveImageException.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
+﻿using System;
+
+namespace MLOps.NET.Exceptions
+{
+    /// <summary>
+    /// Custom exception for failure to run docker image rm
```
Comment (Brett-Parker) [inline src/MLOps.NET/Exceptions/DockerRemoveImageException.cs:6]: Expand on rm

**Review by Brett-Parker — APPROVED** [review]
@aslotte Very minor comments. Happy to merge if you think the comments are not relevant.

**Comment by aslotte** [comment]
Thanks @Brett-Parker! `docker image rm` is the command used to remove an image. Great comment and I expanded on the comment to clarify that

**Review by Brett-Parker — APPROVED** [review]

**Pull request merged**